### DISCUSSION TITLE
Lilygo Lily Pi support

### DISCUSSION THREAD

**Discussion by u4mzu4** [Bodmer/TFT_eSPI#956] [Compatible displays and setup files]
Hello!
First of all I would like you say thanks for the nice library. I'm on the way to order a Lilygo Lily Pi (there are two versions: 3.5" ST7796 and ILI9481). 
http://www.lilygo.cn/prod_view.aspx?TypeId=50032&Id=1353

I would rather order the ILI9481 version, the pinout is here:
https://0.rc.xiniu.com/g3/M00/83/FA/CgAH6F-PtumAKBEXAAGIbootJzA136.jpg

As I recognized, there is a connected TFT_BL pin, where I can control the backlight pith PWM.
If I understood correctly, if I will modify the setup header files properly, there is no big issue expected with this display, am I right?
Thanks in advanced!

**Comment by Bodmer** [comment]
It looks to be designed for the ESP32 but there is limited technical documentation. The TFT_eSPI library does not support capacitive screens but you may find a support library that does.

If you do get one and find it works then do post back as I am sure others will be interested.

**Comment by u4mzu4** [comment]
I'm back with my new toy, based on ILI9481 (hopefully).
I've started with a basic minimal setup, see below:

```
#include <TFT_Drivers/ILI9481_Defines.h>

#define ILI9481_DRIVER
#define TFT_DRIVER 0x9481

#define TFT_MISO 23
#define TFT_MOSI 19
#define TFT_SCLK 18
#define TFT_CS   5   // Chip select control pin
#define TFT_DC   27  // Data Command control pin
#define TFT_RST  -1  // Set TFT_RST to -1 if display RESET is connected to ESP32 board RST
#define TFT_BL   12
#define TFT_BACKLIGHT_ON HIGH

#define SMOOTH_FONT
#define SPI_FREQUENCY  	26666666
#define SPI_READ_FREQUENCY  20000000

```
The examples seem to work, but there are random colorful noisy stripes on the screen. Looks like a good square (~320x320), and a crazy area (~160x320). I've tried all the six different init (ILI9481_INIT_X), but there is no difference, also no effect of the SPI frequency change.
Sorry, I can not describe better, but does it ring a bell for you? Or should I upload some display pictures?

**Reply by Bodmer** [reply]
That sound like the SPI clock rate is too high for the display, the datasheet figure is 10MHz maximum.

I think the ILI9481 uses older silicon technology and cannot be overclocked as much as other displays. I expect 16MHz will work OK, maybe even 20MHz. The read frequency will most likely also need reducing if you are reading from the display.

Thanks for reporting this.

The ST7796 display has a maximum clock speed of 16MHz but works at 40MHz. One of mine works fine at 80MHz but this may be unusual.

**Reply by Bodmer** [reply]
A user in #968 found the maximum was 15MHz.

**Reply by u4mzu4** [reply]
I've checked the original code from Liligo, and they use 26 MHz properly, I've alsoe checked below 10 Mhz, the same issue happened.
In the meantime I've checked the init opportinities, and working properly with 16 bit.

I've modified this:
```
#if defined (TFT_PARALLEL_8_BIT) || defined (RPI_DISPLAY_TYPE)
      writedata(0x55);           // 16 bit colour interface
    #else
//      writedata(0x66);
      writedata(0x55);          // unpretentious workaround to force 16 bit
    #endif
```
I also tried before the #define RPI_DISPLAY_TYPE, but only black screen came, nothing else, maybe there are other details are also affected.
What will be the proper solution?

**Reply by Bodmer** [reply]
I think you need to get in touch with the vendor, I expect Lilygo created a library for it or can give you advice. They may have created a special build of TFT_eSPI.

**Comment by Bodmer** [comment]
Make sure you are using the latest version of TFT_eSPI and [have these lines](https://github.com/Bodmer/TFT_eSPI/blob/master/User_Setup_Select.h#L118-L123) in the library User_Setup_Select.h file.

**Reply by u4mzu4** [reply]
OMG, that was the only missing line, you're right! Thank you for your kind and quick support!
Anyway, for me evertyhing works porperly, do you want any "official" support for LilyGo LilyPi? Should I generate a new setup file, or are we just happy and done now?

**Reply by Bodmer** [reply]
A new setup file would be great. Glad it is working now.

**Comment by u4mzu4** [comment]
I put the new setup file (named as _Setup51_LilyPi_ILI9481.h_) and the modified _User_Setup_Select.h_ on my Github repo (I was not sure if a PR would be a good idea or not).
Here is the link, feel free to modifiy or just takeover:
https://github.com/u4mzu4/LilyGo_LilyPi

(Comment for SPI freq: as we discussed above, the 10 MHz would be the absolute safe side, however the vendor uses these settings, and working for me also properly both read and write).
